src/drivers/usb/hc/ehci_hcd.c

```diff
@@ -576,50 +603,77 @@ static void ehci_qh_fill(struct ehci_hcd *ehci, struct usb_request *req, struct
 static int ehci_request(struct usb_request *req) {
 	struct ehci_hcd *ehci;
 	struct ehci_qh *new_qh;
-
+	struct ehci_qtd_hw *qtd, *qtd_prev, *qtd_first;
 	uint32_t token;
-	int cnt = 0;
 
 	assert(req);
 
 	log_debug("token %d req->len %d", req->token, req->len);
 
 	ehci = hcd_to_ehci(req->endp->dev->hcd);
 
-	token = 0;
-	if (req->token & USB_TOKEN_SETUP) {
+	switch (req->endp->type) {
+	case USB_COMM_CONTROL:
+		/* Setup stage */
+		qtd = ehci_qtd_alloc(ehci);
+		qtd_first = qtd;
 		token = EHCI_QTD_PID_SETUP << 8;
-		cnt++;
-	}
-	if (req->token & USB_TOKEN_IN) {
-		/* Use bufferRounding = 1 to read less than was requested if needed. */
-		token = EHCI_QTD_PID_IN << 8;
-		cnt++;
-	}
-	if ((req->token & USB_TOKEN_OUT) && !(req->token & USB_TOKEN_SETUP)) {
-		token = EHCI_QTD_PID_OUT << 8;
-		cnt++;
-	}
-
-	assert(cnt == 1);
-
-	{
-		struct ehci_qtd_hw *qtd;
-
 		token |= EHCI_QTD_STS_ACTIVE | (3 << 10);
+		ehci_qtd_fill(ehci, qtd, (uintptr_t)&req->ctrl_header,
+			sizeof req->ctrl_header, token, 0);
+		qtd_prev = qtd;
 		qtd = ehci_qtd_alloc(ehci);
-		ehci_qtd_fill(ehci, qtd, (uintptr_t)req->buf, req->len , token, 0);
-
-		if (req->endp->hci_specific) {
-			new_qh = req->endp->hci_specific;
+		/* Data stage (optional) */
+		if (req->len > 0) {
+			if (req->token & USB_TOKEN_OUT)  {
+				token = EHCI_QTD_PID_OUT << 8;
+			} else {
+				token = EHCI_QTD_PID_IN << 8;
+			}
+			token |= EHCI_QTD_STS_ACTIVE | (3 << 10);
```

Comment:
I think it's better to replace `3 << 10` with some macro

Reply:
I added `3 << EHCI_QTD_CERR_SHIFT` instead.